### PR TITLE
Dataset cleanup, gnorm clipping/reporting

### PR DESCRIPTION
**Pull request by daviswer**
Requested (and one extra) additions:

1. Remove deprecated modules from (very large) dataset source file
2. Add support for gradient clipping via config flag, and use it to track progression of gnorm throughout training
3. ~~Impose original Llama init scheme via a post-init function call after model construction~~ (see PR 177 in fms repo)

Remaining todo for future PR: unit tests for dataloaders from closedFMS

### CODE REVIEW

**Comment by lchu-ibm** [comment]
we should keep unit test (your todo) in a different PR.

File: main_training.py
Comment (lchu-ibm) [inline main_training.py:63]: can we add a flag in the config like `use_original_llama_init`?  and default it to True

With this flag name, it also helps users to know what this line does.

Setting to False will use the "potentially superior" init that fms set, as it evolves in the future.

Comment (nairbv) [reply main_training.py:63]: is this a temporary workaround? I'm still not sure what the reasoning is for the current init scheme in the main repo, and suspect we're going to want to make "original" be the standard/default one.

Comment (daviswer) [reply main_training.py:63]: Sure, makes sense

Comment (daviswer) [reply main_training.py:63]: The init scheme in the current repo is more principled than the one here, which is extremely ad-hoc (a hardcoded constant initialization scale of 0.02). But in practice our formula doesn't drift too far from 0.02 for Llama 7b scale, and that init scheme does reflect the actual Llama2 model. So we probably should put 0.02 as the default init in openFMS, and put the more principled / auto-scaling one here as an optional setting

**Comment by daviswer** [comment]
Branch has been updated to remove all the extra stuff surrounding model init. We'll now rely on the updated init scheme in PR 177 to ensure the init scheme matches the original Llama2.

File: pretraining/utils/train_utils.py
Comment (nairbv) [inline pretraining/utils/train_utils.py:47]: Does this function exist on module too? I think it's usually `clip_grad_norm_(model.parameters(), threshold)` ?

Also, this isn't just getting the norm but clipping the gradients in-place too right?The PR description makes me think this was intending to be a tracking-only change.

Comment (daviswer) [reply pretraining/utils/train_utils.py:47]: Yeah it's usually an outside function, but because model is FSDP, FSDP has its own implementation (since clipping of sharded grads requires gradient communication). I added grad clipping because we should have been doing that to begin with - for monitoring only, the user can simply set the clipping threshold arbitrarily high (or we could just default it that way)

Comment (nairbv) [reply pretraining/utils/train_utils.py:47]: I see... so we would set the threshold to something like float('inf') to reproduce how our original model was trained?

Comment (daviswer) [reply pretraining/utils/train_utils.py:47]: Exactly, and that's true, I'll update the title/description. Let me take a look at the conflicts...

**Review by nairbv — APPROVED** [review]
not sure if @lchu-ibm would have any view on grad norm change but LGTM

**Comment by lchu-ibm** [comment]
LG.

@daviswer from the discussion above I assume the dependency PR is https://github.com/foundation-model-stack/foundation-model-stack/pull/177, so we should hold merging until that one is merged?

**Comment by nairbv** [comment]
we'll need to fix the mypy test failures before landing this, but I think the init changes in the main repo can be landed independently / either order

**Comment by lchu-ibm** [comment]
Looks like there are some check failures:

```bash
pretraining/utils/config_utils.py:1: error: Skipping analyzing "fms.models.llama": module is installed, but missing library stubs or py.typed marker  [import-untyped]
pretraining/policies/wrapping.py:3: error: Skipping analyzing "fms.models.llama": module is installed, but missing library stubs or py.typed marker  [import-untyped]
pretraining/policies/ac_handler.py:3: error: Skipping analyzing "fms.models.llama": module is installed, but missing library stubs or py.typed marker  [import-untyped]
pretraining/policies/ac_handler.py:3: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
Found 3 errors in 3 files (checked 13 source files)
Error: Process completed with exit code 1.
```

which seems to be coming from FMS repo, rather than this PR. Now the question would be:

@afrittoli  Do we want to force the per-import check, or we are ok to suppress it?
If we want to enforce this, @JRosenkranz , what's the amount of effort (i.e. the current typed coverage) needed in FMS repo?

**Comment by afrittoli** [comment]
> Looks like there are some check failures:
> 
> ```shell
> pretraining/utils/config_utils.py:1: error: Skipping analyzing "fms.models.llama": module is installed, but missing library stubs or py.typed marker  [import-untyped]
> pretraining/policies/wrapping.py:3: error: Skipping analyzing "fms.models.llama": module is installed, but missing library stubs or py.typed marker  [import-untyped]
> pretraining/policies/ac_handler.py:3: error: Skipping analyzing "fms.models.llama": module is installed, but missing library stubs or py.typed marker  [import-untyped]
> pretraining/policies/ac_handler.py:3: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
> Found 3 errors in 3 files (checked 13 source files)
> Error: Process completed with exit code 1.
> ```
> 
> which seems to be coming from FMS repo, rather than this PR. Now the question would be:
> 
> @afrittoli Do we want to force the per-import check, or we are ok to suppress it? If we want to enforce this, @JRosenkranz , what's the amount of effort (i.e. the current typed coverage) needed in FMS repo?

`mypy` needs the types from imports to check the types in the current package. The FMS repo is already typed (except for the `hf` package where the typing is incomplete), and I added a `py.typed` there, but it seems that it's not picked up for some reason, I will check what's going on. 

> 
> cc @nairbv

**Comment by afrittoli** [comment]
> > Looks like there are some check failures:
> > ```shell
> > pretraining/utils/config_utils.py:1: error: Skipping analyzing "fms.models.llama": module is installed, but missing library stubs or py.typed marker  [import-untyped]
> > pretraining/policies/wrapping.py:3: error: Skipping analyzing "fms.models.llama": module is installed, but missing library stubs or py.typed marker  [import-untyped]
> > pretraining/policies/ac_handler.py:3: error: Skipping analyzing "fms.models.llama": module is installed, but missing library stubs or py.typed marker  [import-untyped]
> > pretraining/policies/ac_handler.py:3: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
> > Found 3 errors in 3 files (checked 13 source files)
> > Error: Process completed with exit code 1.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > which seems to be coming from FMS repo, rather than this PR. Now the question would be:
> > @afrittoli Do we want to force the per-import check, or we are ok to suppress it? If we want to enforce this, @JRosenkranz , what's the amount of effort (i.e. the current typed coverage) needed in FMS repo?
> 
> `mypy` needs the types from imports to check the types in the current package. The FMS repo is already typed (except for the `hf` package where the typing is incomplete), and I added a `py.typed` there, but it seems that it's not picked up for some reason, I will check what's going on.
> 
> > cc @nairbv



> > Looks like there are some check failures:
> > ```shell
> > pretraining/utils/config_utils.py:1: error: Skipping analyzing "fms.models.llama": module is installed, but missing library stubs or py.typed marker  [import-untyped]
> > pretraining/policies/wrapping.py:3: error: Skipping analyzing "fms.models.llama": module is installed, but missing library stubs or py.typed marker  [import-untyped]
> > pretraining/policies/ac_handler.py:3: error: Skipping analyzing "fms.models.llama": module is installed, but missing library stubs or py.typed marker  [import-untyped]
> > pretraining/policies/ac_handler.py:3: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
> > Found 3 errors in 3 files (checked 13 source files)
> > Error: Process completed with exit code 1.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > which seems to be coming from FMS repo, rather than this PR. Now the question would be:
> > @afrittoli Do we want to force the per-import check, or we are ok to suppress it? If we want to enforce this, @JRosenkranz , what's the amount of effort (i.e. the current typed coverage) needed in FMS repo?
> 
> `mypy` needs the types from imports to check the types in the current package. The FMS repo is already typed (except for the `hf` package where the typing is incomplete), and I added a `py.typed` there, but it seems that it's not picked up for some reason, I will check what's going on.

Oh, it's the install command, it looks like it must be installed in dev mode from the a specific sha, otherwise it picks v0.0.3, which does not contain the `py.typed` command. I will make a PR to fix it now.

**Comment by lchu-ibm** [comment]
re-running type checking as https://github.com/foundation-model-stack/fms-fsdp/pull/10 has been merged.

I will merge this PR once that passes.

**Comment by lchu-ibm** [comment]
@daviswer @nairbv 
I think this PR does require https://github.com/foundation-model-stack/foundation-model-stack/pull/177 to be landed first due to the introduction of `reset_parameters`.

I will try to push on the other end otherwise I might have to revert this merge as it is broken.

**Comment by nairbv** [comment]
@lchu-ibm what's the failure case you're seeing?

I think in this commit the `model.reset_parameters()` should have been gated on `hasattr`:

https://github.com/foundation-model-stack/fms-fsdp/pull/5/commits/8a275873160ae5620471747cc0d08053dd258146

That commit came after my comment, but not sure if that's the breakage you're seeing.

**Comment by nairbv** [comment]
ah I see this issue: https://github.com/foundation-model-stack/fms-fsdp/issues/15

that makes sense. model needs to be initialized before applying fsdp.

**Pull request merged**
